### PR TITLE
AST changes for channel types

### PR DESCRIPTION
**Pull request by eyingxuan**

### CODE REVIEW

**Comment by KeenWill** [comment]
oh whoops i already started this but forgot to push

**Comment by KeenWill** [comment]
I'll see if there's anything I can merge in

File: lib/ast/ast.ml [hunk 1]
```diff
@@ -8,7 +8,14 @@ let no_loc x = { elt = x; loc = Range.norange }
...
+type ty = TRegTy of regty | TLinTy of lty
+
+and lty = TChan of ty * mult * mult
+
+(* multiplicity of usage: 0, 1, arbitrary *)
+and mult = MNum of int64 | MArb
+
+and regty = TBool | TInt | TRef of rty | TNullRef of rty
 
 and rty = RString | RStruct of id | RArray of ty | RFun of ty list * ret_ty
```
Comment (eyingxuan) [inline lib/ast/ast.ml:21]: Couldn't figure out how to factor out `regty` and make this `RArray of regty` instead in the parser without shift-reduce conflicts. I'm guessing we have to do this in the typechecking stage and make sure all arrays/structs don't have channels in them.

**Comment by eyingxuan** [comment]
> oh whoops i already started this but forgot to push

Oh crap, sorry!! I totally should've asked

**Pull request closed without merging**